nnvm/python/nnvm/frontend/tensorflow.py

```diff
@@ -1177,43 +1183,65 @@ def from_tensorflow(self, graph, layout="NHWC", shape=None, outputs=None):
             raise NotImplementedError( \
                 "The following operators are not implemented: {}".format(missing_operators))
 
+        for node in graph.node:
+            if node.op == 'Placeholder':
+                self._input_shapes[node.name] = \
+                    tensor_util.TensorShapeProtoToList(node.attr['shape'].shape)
+
+                if shape and node.name in shape:
+                    for idx, dim in enumerate(self._input_shapes[node.name]):
+                        if dim < 0:
+                            self._input_shapes[node.name][idx] = shape[node.name][idx]
+                    if not self._input_shapes[node.name]:
+                        self._input_shapes[node.name] = list(shape[node.name])
+                    assert self._input_shapes[node.name] == list(shape[node.name])
+                else:
+                    for idx, dim in enumerate(self._input_shapes[node.name]):
+                        if dim < 0:
+                            self._input_shapes[node.name][idx] = 1
```

Comment:
is it true in definition? just want to confirm.

Reply:
Thanks for pointing this out! For some of the tf graphdefs, they might have -1 in input placeholder, here -1 is the value of N in NCHW or other layouts, it will cause errors like the following when inferring shape using graph_util. For this case, we set 1 as default since tvm hasn't supported dynamic batch so far as I know. 

```
 File ".../lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/frontend/tensorflow.py", line 1302, in from_tensorflow
    list(graph_util.infer_shape(g, **self._input_shapes))[-1]
  File "...l/lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/compiler/graph_util.py", line 31, in infer_shape
    graph = graph.apply("InferShape")
  File ".../lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/graph.py", line 234, in apply
    check_call(_LIB.NNGraphApplyPasses(self.handle, npass, cpass, ctypes.byref(ghandle)))
  File ".../lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/_base.py", line 75, in check_call
    raise NNVMError(py_str(_LIB.NNGetLastError()))
nnvm._base.NNVMError: [15:19:47] 
.../TVM/tvm/nnvm/src/top/nn/convolution.cc:120: Operator conv2d(kernel_layout=HWIO, use_bias=False, channels=64, kernel_size=(7, 7), strides=(2, 2), padding=[0, 0], layout=NHWC, dilation=(1, 1), name=resnet_v1_50/conv1/Conv2D) expects data's shape to be [4294967295,230,230,3], but got [-1,230,230,3].
```

Reply:
It need not be ```-1``` for only N but for other attributes too.
I suggest to assert simply and force user to supply the complete shape.

Reply:
Ok, print the warning info indicating this assumption.